ironic/drivers/modules/drac/management.py

```diff
@@ -302,14 +313,173 @@ def set_boot_device(node, device, persistent=False):
+        :returns: None if it is completed.
+        :raises: RedfishError on an error.
+        """
+        system = redfish_utils.get_system(task.node)
+        for manager in system.managers:
+            try:
+                oem_manager = manager.get_oem_extension('Dell')
+            except sushy.exceptions.OEMExtensionNotFoundError as e:
+                error_msg = (_("Search for Sushy OEM extension package "
+                           "'sushy-oem-idrac' failed for node %(node)s. "
+                           "Ensure it is installed. Error: %(error)s") %
+                         {'node': task.node.uuid, 'error': e})
+                LOG.error(error_msg)
+                raise exception.RedfishError(error=error_msg)
+            try:
+                reset_job_response = oem_manager.reset_idrac(task)
+            except sushy.exceptions.SushyError as e:
+                error_msg = ("Sushy OEM extension Python package "
+                          "sushy-oem-idrac failed to reset idrac"
+                          "for %(node)s, Error : %(error)s" %
+                          {'node':task.node.uuid, 'error':e})
+                LOG.error(error_msg)
+                raise exception.RedfishError(error=error_msg)
+
+        redfish_address = task.node.driver_info.get('redfish_address')
+        redfish_address = urlparse(redfish_address)
+        redfish_hostname = redfish_address.netloc
+
+        if reset_job_response.status_code == _JOB_RESPONSE_CODE:
+            msg = ("iDRAC was reset for node %(node)s via OEM,"
+                    "waiting for return to operational state" %
+                    {'node':task.node.uuid})
+            LOG.info(msg)
+            LOG.debug(msg)
+            redfish_utils.wait_for_host(redfish_hostname)
+            redfish_utils.wait_until_idrac_is_ready(oem_manager,
+                        redfish_hostname,
+                        IDRAC_IS_READY_RETRIES,
+                        IDRAC_IS_READY_RETRY_DELAY_SEC)
+        else:
+            error_msg = ("Failed to reset idrac for %(node)s" %
+                        {'node': task.node.uuid})
+            LOG.error(error_msg)
+            raise exception.RedfishError(error=error_msg)
+
+        return None
+
+    @METRICS.timer('DracRedfishManagement.known_good_state')
+    @base.clean_step(priority=0)
+    def known_good_state(self, task):
+        """Reset the iDRAC, Clear the job queue.
+
+        :param task: a TaskManager instance containing the node to act on.
+        :returns: None if it is completed.
+        :raises: RedfishError on an error.
+        """
+        self.reset_idrac(task)
+        self.clear_job_queue(task)
+        return None
+
+    @task_manager.require_exclusive_lock
+    def set_boot_device(self, task, device, persistent=False):
+        super(
+            DracRedfishManagement,
+            self).set_boot_device(
+            task,
+            device,
+            persistent=False)
```

Comment:
The call to set_boot_device causes a config job to be created under the covers.  We need to make sure that there are no pending config jobs prior to setting the boot device, otherwise setting the boot device will fail with an error like "a pending config job already exists".
As a result, we need to move this call to super.set_boot_device to after the "if" statement below.  The test to see if this is working is to create a pending config job, and then try running one of these new cleaning steps.  Ironic will call set_boot_device to set the node to PXE boot, and boot up IPA.  After that, it will run the cleaning step.  If the pending job is not deleted then the attempt to set the boot device will fail.

Reply:
Yes Chris, we should move set_boot_device call after deleting pending jobs. Addressed it.